test.js

```diff
@@ -546,7 +546,12 @@ test('should support a single spread prop', () => {
     callee: {type: 'Identifier', name: 'h'},
     arguments: [
       {type: 'Literal', value: 'a'},
-      {type: 'Identifier', name: 'b'}
+      {
+        type: 'ObjectExpression',
+        properties: [
+          {type: 'SpreadElement', argument: {type: 'Identifier', name: 'b'}}
+        ]
+      }
```

Comment:
Note that this test is about the classic runtime.

Old behaviour: Pass a single JSX spread element by reference\
New behavour: Always pass a shallow clone, no exception for a single JSX spread element.

The old behaviour is was consistent with Babel for the classic runtime. The new behaviour is consistent with Babel for the automatic runtime, as well as TypeScript for both the classic and the automatic runtime.

I like the new behaviour, but I’m open to changing it.

Reply:
It depends.

I think React creates shallow clones internally anyway, so it’s safe. For other frameworks it may be unsafe to pass objects by reference. However, I haven’t seen any issues with this in practice, which makes me think it’s probably fine to pass by reference.

I also suspect JavaScript engines optimize this anyway, but this is just a guess.

Another difference is that the spread syntax filters own properties.

I think it would be best to keep the new spread syntax, but I’ll gladly change it if you have strong feelings about passing by reference.

Reply:
IMO this isn’t about performance. It’s if we want to make a special case of passing one single JSX spread.

```jsx
/** @jsx h */
class Person {
  constructor(firstName, lastName) {
    this.firstName = firstName
    this.lastName = lastName
  }

  getFullName() {
    return this.firstName + ' ' + this.lastName
  }
}

function h(tagName, props) {
  console.log(props.getFullName())
}

// Logs 'Spongebob Squarepants'
<tag {...new Person('Spongebob', 'Squarepants')} />

// Uncaught TypeError: props.getFullName is not a function
<tag {...new Person('Spongebob', 'Squarepants')} additional="prop" />
```